Refactored CallableResolver into a stateless service focused on resolving callables only

mnapoli opened the pull request:
Per discussions on IRC, I refactored CallableResolver into a stateless service focused on resolving callables only (not wrapping and invoking them).

This will lead to less object instantiations (since the service is now stateless instead of wrapping every middleware/controller callable), as well as better separation of concerns since the class' responsibility is now only to resolve callables. It isn't responsible of invoking callables anymore (responsibility which was spread between here and also the invokation strategy and middleware stack too IIRC).

One detail is that I have moved the `$callable->bindTo($container)` calls into the CallableResolver itself since it belongs there IMO (because it's about preparing callables for being invokable correctly). That avoids having "callable resolution" code all over the place (App and Route).

Feedback welcome. I'll probably be able to answer and update the PR in 2 days.

mnapoli commented on tests/RouteTest.php:
```diff
@@ -153,6 +154,8 @@ public function testAddMiddlewareAsString()
 
         $container = new Container();
+        $container['MiddlewareStub'] = new MiddlewareStub();
+
```
Could someone explain to me how it was working before? I'm afraid I introduced a regression because now I have to register the middleware in the container (and it's not necessary on 3.x).

JoeBengalen replied:
It worked before because the `CallableResolver` was never really invoked, so it did not try to find the `MiddlewareStub`.

mnapoli replied:
OK so that's an improvement then ;)

mnapoli commented:
Also the next step IMO is to remove the `CallableResolverAwareTrait` (again, per discussions on IRC), but I wanted to do it in a separate pull request.

JoeBengalen commented on Slim/CallableResolver.php:
```diff
@@ -25,87 +24,58 @@
-    }
-
     /**
      * Resolve toResolve into a closure that that the router can dispatch.
```
Comment is outdated

mnapoli replied:
The new signature of the method is `public function resolve($toResolve)` so it makes sense right?

JoeBengalen commented:
Maybe a final check should happen in `CallableResolver::resolve()` if `$resolved` is indeed callable. 

Currently, anything you pass that is not a string or callable will be returned without modification. And since the `resolve()` claims to return a callable at all times I think an Exception should be thrown if the resulting `$resolved` is not callable.

mnapoli commented:
Thanks for the review!

> Maybe a final check should happen in CallableResolver::resolve() if $resolved is indeed callable.

Done in ba08a05, an exception will be thrown if the CallableResolver cannot return a real callable.

> Closures will never be bound to the container 

Oh yeah good catch, I fixed it in e87d996. I also added a test for that to make sure it's staying.

JoeBengalen commented:
@ppetermann, good question ... It didn't change in this PR but was already bound to container instead of app. No sure where and why it was changed.

All given examples in de docs assume closures are indeed bound to app.

juliangut commented:
`CallableResolver` is used to resolve both middlewares and routes, so either bindTo is moved out from CallableResolver or it needs a parameter to be told what to bind to?

mnapoli commented:
Thank you for noticing that. I've restored the way it was before regarding binding Closures to app or container.

akrabat commented:
Looks good to me.

akrabat commented:
Thanks @mnapoli
